gammapy/cube/ring.py

```diff
@@ -131,122 +110,139 @@ def kernels(self, image):
-        counts, exposure_on, exclusion = [images[_] for _ in required]
-
-        if not counts.geom.is_image:
-            raise ValueError("Only 2D maps are supported")
 
+        counts = dataset.counts
+        background = dataset.background_model.map
         kernels = self.kernels(counts)
-        cubes = {
-            "exposure_on": self._exposure_on_cube(exposure_on, kernels),
-            "exposure_off": self._exposure_off_cube(exposure_on, exclusion, kernels),
-            "off": self._off_cube(counts, exclusion, kernels),
-        }
-        cubes["alpha_approx"] = self._alpha_approx_cube(cubes)
 
-        exposure_off, off = self._reduce_cubes(cubes)
-        alpha = exposure_on.data / exposure_off
+        cubes = {}
+        cubes["counts_off"] = scale_cube(
+            (counts.data * exclusion.data)[0, Ellipsis], kernels
+        )
+        cubes["acceptance_off"] = scale_cube(
+            (background.data * exclusion.data)[0, Ellipsis], kernels
+        )
+
+        scale = background.geom.pixel_scales[0].to("deg")
+        theta = self.parameters["theta"] * scale
+        tophat = Tophat2DKernel(theta.value)
+        tophat.normalize("peak")
+        acceptance = background.convolve(tophat.array)
+        acceptance_data = acceptance.data[0, Ellipsis]
+        cubes["acceptance"] = np.repeat(
+            acceptance_data[Ellipsis, np.newaxis], len(kernels), axis=2
+        )
+
+        return cubes
 
-        # set data outside fov to zero
-        not_has_exposure = ~(exposure_on.data > 0)
-        for data in [alpha, off, exposure_off]:
-            data[not_has_exposure] = 0
+    def run(self, dataset, exclusion):
+        """Run adaptive ring background maker
 
-        background = alpha * off
+        Parameters
+        ----------
+        dataset : `~gammapy.cube.fit.MapDataset`
+            Input map dataset.
+        exclusion : `~gammapy.maps.WcsNDMap`
+            Exclusion mask for regions with known gamma-ray emission.
 
-        return {
-            "exposure_off": counts.copy(data=exposure_off),
-            "off": counts.copy(data=off),
-            "alpha": counts.copy(data=alpha),
-            "background_ring": counts.copy(data=background),
-        }
+        Returns
+        -------
+        dataset_on_off : `~gammapy.cube.fit.MapDatasetOnOff`
+            On off dataset.
+        """
+        cubes = self.make_cubes(dataset, exclusion)
+        acceptance, acceptance_off, counts_off = self._reduce_cubes(cubes, dataset)
 
+        not_has_off_acceptance = acceptance_off.data <= 0
+        acceptance_off.data[not_has_off_acceptance] = np.nan
 
-class RingBackgroundEstimator:
+        fft_noise_threshold = 1e-6
+        not_has_acceptance = acceptance.data <= fft_noise_threshold
+        counts_off.data[not_has_acceptance] = 0
+        acceptance_off.data[not_has_acceptance] = 0
+        acceptance.data[not_has_acceptance] = 0
+
+        mask_safe = dataset.counts.copy(
```

Comment:
I think we should definitely keep and use the `mask_safe`, probably instead of modifying the data, as you did above. My naive proposal would be, that the `mask_safe` should be taken into account when convolving the acceptance (instead of relying on it to be zero) and then the `mask_safe` should be modified to exclude regions with no `acceptance_off` (computed as I proposed above). 

I have to think about it once more, to get a clearer picture of how to handle the `mask_safe` here, but I think in general we should leave the data untouched (not setting anything to zero or `NaN`) but keep track of the information in the `mask_safe` instead. When the dataset is stacked the data is modified anyway (set to zero). And this gives a single, well defined place, where the data is touched.

As I said: I'll think about it once more and make a proposal. It should not stop us form merging this PR.